jetstream_pt/quantize.py

```diff
@@ -95,3 +97,63 @@ def load_q_weight_helper(w_q, scale, zp=None, block_size=-1):
       zp = (zp * scale).transpose(1, 0).squeeze(-1).to(torch.bfloat16)
     scale = scale.transpose(1, 0).squeeze(-1).to(torch.bfloat16)
   return w_q, scale, zp
+
+
+def blockwise_jax_kernel(inputs, weight, weight_scaler, zero_point):
+  """Blockwise Matmul kernel impl in JAX using einsum"""
+  weight = weight.astype(jnp.int8)
+  block_size = weight.shape[1]
+  inputs_shape = inputs.shape
+  inputs_new_shape = inputs_shape[:-1] + (
+      inputs_shape[-1] // block_size,
+      block_size,
+  )
+  inputs = inputs.reshape(inputs_new_shape)
+  out = jnp.einsum("scz,bdsc->bdsz", weight, inputs)
+  out = jnp.einsum("bdsz,sz->bdz", out, weight_scaler)
```

Comment:
Because scaler has shape `[IN_DIM / BLOCK_SIZE, OUT_DIM]`. This is because with blockwise quant, every `BLOCK_SIZE` elements share a bf16 scaler.

Other Matmul inputs: `x: [BS, SEQ, IN_DIM]`, `w: [IN_DIM / BLOCK_SIZE, BLOCK_SIZE, OUT_DIM]` (`w` has shape `[IN_DIM, OUT_DIM]` for bf16 weight)

The reduction operation is split into 2 einsum since the reduction axis is split into 2.


With the above inputs, the matmul becomes:

1. Reshape `x` to `[BS, SEQ, IN_DIM / BLOCK_SIZE, BLOCK_SIZE]`. The contract dim is split into `IN_DIM / BLOCK_SIZE` and `BLOCK_SIZE`.
2. Reduce along the `BLOCK_SIZE` dim: `x: [BS, SEQ, IN_DIM / BLOCK_SIZE, BLOCK_SIZE] @ w: [IN_DIM / BLOCK_SIZE, BLOCK_SIZE, OUT_DIM] => matmul_out [BS, SEQ, IN_DIM / BLOCK_SIZE, OUT_DIM]`
3. Reduce along `IN_DIM / BLOCK_SIZE` dim `matmul_out [BS, SEQ, IN_DIM / BLOCK_SIZE, OUT_DIM] @ scaler: [IN_DIM / BLOCK_SIZE, OUT_DIM] => matmul_out [BS, SEQ, OUT_DIM]`

Reply:
thanks, this is very clear.